File: 29-divide-two-integers/divide-two-integers.cpp
class Solution {
public:
    int divide(int dividend, int divisor) {
        bool isDividendNeg = false, isDivisorNeg = false;
        long long divid = dividend, divis = divisor;

        if (divis == 1){
            return divid;
        } else if (divis == -1){
            if (divid > 0) {
                return -divid;
            } else {
                if (-divid > INT_MAX) return INT_MAX;
                else return -divid;
            }
        }
        
        if (dividend < 0) {
            isDividendNeg = true;
            divid *= -1;
        }

        if (divisor < 0) {
            isDivisorNeg = true;
            divis *= -1;
        }

        long long quotient = 0;
        long long d = divis;
        while (d <= divid) {
            quotient += 1;
            d += divis;
        }

        if ((isDividendNeg == true && isDivisorNeg == false) ||
            (isDividendNeg == false && isDivisorNeg == true)) {
            quotient *= -1;
        }

        if (quotient > INT_MAX)
            return INT_MAX;
        else if (quotient < INT_MIN)
            return INT_MIN;

        return quotient;
    }
};

File: 29-divide-two-integers/divide-two-integers.go
func divide(dividend int, divisor int) int {
    if dividend == 0 {
        return 0
    }

    if divisor == 1 {
        return dividend
    } else if divisor == -1 {
        quo := -1 * dividend
        fmt.Println(quo)
        if quo > math.MaxInt32 {
            return math.MaxInt32
        }
        return quo
    }

    dividend_neg, divisor_neg := false, false
    if dividend <  0 {
        dividend *= -1
        dividend_neg = true
    } 
    if divisor < 0 {
        divisor *= -1
        divisor_neg = true
    }

    sum := divisor
    quo := 0
    for dividend >= sum {
        quo++
        sum += divisor
    }

    if (dividend_neg == true && divisor_neg == false) || (dividend_neg == false && divisor_neg == true) {
        return -1 * quo
    }

    return quo
}